### PR TITLE
core.meta: Implemented documentation generator for *.melk files #91

### PR DESCRIPTION
**Pull request by DAGobertGans**
The MelkDocumentationGenerator creates a Markdown file for each algorithm, group and option in a bundle.
These files contain documentation like descriptions, identifiers, supported options, default values, dependencies etc.
Furthermore it is possible to specify additional documentation for any bundle member via a 'documentation' field.
The generated files reside in the the directory 'docs' of the elk repository. GitHub utilizes the static-site-generator Jekyll
to compile the Markdown files into a website.

### CODE REVIEW

File: plugins/org.eclipse.elk.core.meta/src/org/eclipse/elk/core/meta/jvmmodel/MelkDocumentationGenerator.xtend [hunk 1]
```diff
@@ -0,0 +1,560 @@
+/*******************************************************************************
+ * Copyright (c) 2016 dag and others.
```
Comment (uruuru) [inline plugins/org.eclipse.elk.core.meta/src/org/eclipse/elk/core/meta/jvmmodel/MelkDocumentationGenerator.xtend]: Kiel University

File: plugins/org.eclipse.elk.core.meta/META-INF/MANIFEST.MF [hunk 2]
```diff
@@ -16,7 +16,8 @@ Require-Bundle: org.eclipse.xtext;bundle-version="2.9.0",
  org.eclipse.xtext.util,
  org.eclipse.xtend.lib,
  org.eclipse.emf.common,
- org.eclipse.elk.graph
+ org.eclipse.elk.graph,
+ org.eclipse.elk.core;bundle-version="0.2.0"
```
Comment (uruuru) [inline plugins/org.eclipse.elk.core.meta/META-INF/MANIFEST.MF]: No version

File: plugins/org.eclipse.elk.core.meta/src/org/eclipse/elk/core/meta/jvmmodel/MelkDocumentationGenerator.xtend [hunk 3]
```diff
@@ -0,0 +1,560 @@
...
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *    dag - initial API and implementation
```
Comment (le-cds) [inline plugins/org.eclipse.elk.core.meta/src/org/eclipse/elk/core/meta/jvmmodel/MelkDocumentationGenerator.xtend]: This should be either "Kiel University" or your name.

**Comment by le-cds** [comment]
Once those small changes are pushed, I'll be happy to merge the pull request. :)

**Comment by uruuru** [comment]
One point we should have at least discussed before we merge this: are we fine with the fact that the documentation, including images, is part of the repository? The repository may become quite large.

File: docs/_elk-wiki/20_30_Using-Plain-Java-Layout.md [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+When running in an Eclipse context, the meta data service is magically populated. In a plain Java application, you will have to populate the service yourself. Layout meta data are provided by `ILayoutMetaDataProvider`s. What you need to do is to obtain an instance of each relevant meta data provider and register it with the meta data service.
+
+One such meta data provider is always automatically registered: `CoreOptions`. This will give you access to our most basic layout algorithms: _Fixed Layout_, _Box Layout_, and _Random Layout_ (at the time of writing). To use more layout algorithms, you will have to find the algorithm's implementation of `ILayoutMetaDataProvider` and register that. For example, using _ELK Layered_ requires the following registration:
+
+```java
+LayoutMetaDataService service = LayoutMetaDataService.getInstance();
+service.registerLayoutMetaDataProvider(new LayeredOptions());
```
Comment (uruuru) [inline docs/_elk-wiki/20_30_Using-Plain-Java-Layout.md]: ``LayeredOptions`` is old, must be `LayeredMetaDataService` now.

**Comment by spoenemann** [comment]
> Are we fine with the fact that the documentation, including images, is part of the repository? The repository may become quite large.

We could request an additional repository into which the docs are generated.

**Comment by le-cds** [comment]
Actually, I don't see much of a problem with leaving the documentation in the main repository. Having to check out another repository would actually cause problems for people that don't have the repository checked out in the directory the generator expects.

Come to think of it, what we could do is add a docs folder to our main repository that points to another repository. I think Git supports that? Don't you Xtext people use that feature? The question is whether or not GitHub's page hosting code supports that...

**Comment by spoenemann** [comment]
>  Don't you Xtext people use that feature?

Never heard of it.

**Comment by eNBeWe** [comment]
Am 30.01.2017 um 09:42 schrieb Christoph Daniel Schulze:
> Come to think of it, what we could do is add a docs folder to our main
> repository that points to another repository. I think Git supports that?

There are at least two ways to do that.

submodule: https://git-scm.com/docs/git-submodule
subtree: (Which I cannot find in the git-scm doc) ...
https://blogs.atlassian.com/2013/05/alternatives-to-git-submodule-git-subtree/

Whether it is supported by github pages, that is another question.

**Comment by le-cds** [comment]
> Never heard of it.

You're right, my recollection seems to have been wrong.

> Whether it is supported by github pages, that is another question.

I'll check that.

**Comment by le-cds** [comment]
It seems to me that using submodules or subtrees on this scale is problematic since using them is not straightforward. Many developers haven't even heard of these features.

I personally think the best solution is still to simply let the documentation generator generate into the main repository and commit changes to the documentation together with the melk file changes they were caused by.

**Pull request merged**